u_http.go

```diff
@@ -0,0 +1,565 @@
+type uhttpStringer struct {
+	uhttpCloseableTransport
+	name string
+}
+
+// String returns a compact string representation of a transport.
+func (uhs *uhttpStringer) String() string {
+	return fmt.Sprintf("%s#%p", uhs.name, uhs)
+}
+
+// maybeInitTxps initializes the internal transports once.
+func (txp *UHTTPTransport) maybeInitTxps() {
+	defer txp.mu.Unlock()
+	txp.mu.Lock()
+	if txp.cleartext == nil {
+		txp.cleartext = &uhttpStringer{
+			uhttpCloseableTransport: &http.Transport{
+				DialContext: txp.dialCleartext,
+				Proxy:       txp.Proxy,
+			},
+			name: "cleartext",
+		}
+	}
+	if txp.onlyDial == nil {
+		txp.onlyDial = &uhttpStringer{
+			uhttpCloseableTransport: &http.Transport{
+				DialContext:    txp.disableDialContext,
+				DialTLSContext: txp.dialUTLSContext,
+			},
+			name: "onlyDial",
+		}
+	}
+	if txp.https == nil {
+		txp.https = &uhttpStringer{
+			uhttpCloseableTransport: &http.Transport{
+				DialTLS: txp.connCacheDialTLSHTTPS,
+			},
+			name: "https",
+		}
+	}
+	if txp.h2 == nil {
+		txp.h2 = &uhttpStringer{
+			uhttpCloseableTransport: &http2.Transport{
+				DialTLS: txp.connCacheDialTLSH2,
+			},
+			name: "h2",
+		}
+	}
+}
+
+// dialCleartext calls DialContext or uses a default DialContext
+// if no DialContext has been configured by the user.
+func (txp *UHTTPTransport) dialCleartext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	dialFn := txp.DialContext
+	if dialFn == nil {
+		dialFn = (&net.Dialer{}).DialContext
+	}
+	uhttpLog.Printf("uhttp: dialCleartext %p %s %s", ctx, network, address)
+	return dialFn(ctx, network, address)
+}
+
+// disableDialContext is a DialContext that always fails.
+func (txp *UHTTPTransport) disableDialContext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	return nil, errors.New("uhttp: DialContext should not have been called")
+}
+
+// connCacheDialTLSHTTPS returns a cached connection for the given
+// address, if any, otherwise errUHTTPNoCachedConn.
+func (txp *UHTTPTransport) connCacheDialTLSHTTPS(
+	network, address string) (net.Conn, error) {
+	if conn := txp.connCachePop(address); conn != nil {
+		return conn, nil
+	}
+	uhttpLog.Printf("uhttp: https: connCache miss for %s", address)
+	return nil, errUHTTPNoCachedConn
+}
+
+// connCacheDialTLSH2 returns a cached connection for the given
+// address, if any, otherwise errUHTTPNoCachedConn.
+func (txp *UHTTPTransport) connCacheDialTLSH2(
+	network, address string, config *tls.Config) (net.Conn, error) {
+	if conn := txp.connCachePop(address); conn != nil {
+		return conn, nil
+	}
+	uhttpLog.Printf("uhttp: h2: connCache miss for %s", address)
+	return nil, errUHTTPNoCachedConn
+}
+
+// dialUTLSContext dials a TLS connection using UTLS and the
+// settings configured inside UHTTPTransport. This function
+// updates hostCache and saves the connection into connCache,
+// on success. Note that success is indicated by returning
+// one of errUHTTPUse{H2,HTTPS}.
```

Comment:
Let's not indicate success by returning errors. Seems like this will need some refactoring.

Reply:
Good point. Side effects and errors as sentinels are bad. Thank you for your comment, this is really what code review is about! I suppose there is here a good opportunity for simplifying the codebase by getting rid of `onlyDial` and running the `dialUTLSContext` right inside `RoundTrip`. By running the diff on memory, this seems possible to do and it also seems to yield a simpler algorithm inside of `RoundTrip`, which is also good.